cpp/react-native-quick-sqlite.cpp

```diff
@@ -306,54 +330,74 @@ void installSequel(jsi::Runtime &rt, const char *docPath)
-  //                                                                                                   {
-  //                                                                                                     jsi::detail::throwJSError(rt, result.message.c_str());
-  //                                                                                                   }
-  //                                                                                                   else
-  //                                                                                                   {
-  //                                                                                                     resolve->asObject(rt).asFunction(rt).call(rt, move(result.value));
-  //                                                                                                   }
-  //                                                                                                 });
-
-  //                                                                                                 t1.detach();
-
-  //                                                                                                 return {};
-  //                                                                                               }));
-
-  //       return promise;
-  //     });
+  auto asyncExecSQL = jsi::Function::createFromHostFunction(
+      rt,
+      jsi::PropNameID::forAscii(rt, "sequel_asyncExecSQL"),
+      3,
+      [](jsi::Runtime &rt, const jsi::Value &thisValue, const jsi::Value *args, size_t count) -> jsi::Value
+      {
+        const string dbName = args[0].asString(rt).utf8(rt);
+        const string query = args[1].asString(rt).utf8(rt);
+        std::shared_ptr<jsi::Value> params = std::make_shared<jsi::Value>(args[2].asObject(rt));
+
+        std::shared_ptr<jsi::Object> resolver;
+
+        auto promiseBody = jsi::Function::createFromHostFunction(
+            rt,
+            jsi::PropNameID::forAscii(rt, "executor"),
+            2,
+            [&resolver](
+                jsi::Runtime &rt,
+                const jsi::Value &thisValue,
+                const jsi::Value *args,
+                size_t) -> jsi::Value
+            {
+              resolver = std::make_shared<jsi::Object>(args[0].asObject(rt));
+              return {};
+            });
+
+        auto promise = rt.global()
+                           .getPropertyAsFunction(rt, "Promise")
+                           .callAsConstructor(rt, promiseBody);
+
+        // Spawn c++ thread
+        thread t1(
+            [&rt, dbName, query, params, resolver]
+            {
+              SequelResult result = sequel_execute(rt, dbName, query, *params);
+
+              if (result.type == SequelResultError)
+              {
+                resolver->asFunction(rt).call(rt, createError(rt, result.message.c_str()));
```

Comment:
Yeah, I've been trying to get fbjni and the callInvoker work in #29 but so far I cannot get android to compile with the jni, looking at your repo, I've tried to replicate the `build.gradle` and the `CMakeLists` but it is way too complex for me

Reply:
i have watched your video on youtube, i know why on android you have a crashes, it because you looked at my code and at logcat errors not very well) to pass java object like jsInvoker to jni you need to rewrite jni communication between java and c++. https://github.com/sergeymild/jsi-sqlite - here i created temporally library there i did everything like it must be on android. And it works as expected. Also you don't need to rewrite sqlite.cpp.

Reply:
currently it's only implemented on android to show you that you did wrong

Reply:
> currently it's only implemented on android to show you that you did wrong

Using your examples, Its working now with this changes:

```
public class NativeProxy {


  private native void installNativeJsi(long jsContextNativePointer, CallInvokerHolderImpl jsCallInvokerHolder, String docDir);
  public static final NativeProxy instance = new NativeProxy();

  public void installJsi(ReactContext context) {
    Log.d("rn-native-quick-sqlite", "Installing native...");
    CallInvokerHolderImpl holder = (CallInvokerHolderImpl)context.getCatalystInstance().getJSCallInvokerHolder();
    long contextPointer = context.getJavaScriptContextHolder().get();
    final String path = context.getFilesDir().getAbsolutePath();
    installNativeJsi(contextPointer, holder, path);
  }
}
```

```
#include <jni.h>
#include <fbjni/fbjni.h>
#include <jsi/jsi.h>
#include <ReactCommon/CallInvokerHolder.h>
#include "react-native-quick-sqlite.h"
#include "logs.h"
#include <typeinfo>

struct NativeProxy : jni::JavaClass<NativeProxy> {
    static constexpr auto kJavaDescriptor = "Lcom/reactnativesequel/NativeProxy;";

    static JavaVM* vm;

    static void registerNatives(JavaVM* v) {
        //NativeProxy::vm = v;
        // register native methods for Java
        javaClassStatic()->registerNatives({
                                                   //initialization for JSI
                                                   makeNativeMethod("installNativeJsi", NativeProxy::installNativeJsi)
                                           });
    }

private:
    static void installNativeJsi(jni::alias_ref<jni::JObject> thiz,
                                 jlong jsiRuntimePtr,
                                 jni::alias_ref<react::CallInvokerHolder::javaobject> jsCallInvokerHolder,
                                 jni::alias_ref<jni::JString> docPath) {

        auto jsiRuntime = reinterpret_cast<jsi::Runtime*>(jsiRuntimePtr);
        auto jsCallInvoker = jsCallInvokerHolder->cthis()->getCallInvoker();
        std::string docPathString = docPath->toStdString();

        installSequel(*jsiRuntime, jsCallInvoker, docPathString.c_str());
    }
};

JNIEXPORT jint JNI_OnLoad(JavaVM* vm, void*) {
    return jni::initialize(vm, [vm] {
        NativeProxy::registerNatives(vm);
    });
}
```

Thank you.

Reply:
@sergeymild Can't access the repo ([sergeymild/react-native-jsi-sqlite](https://github.com/sergeymild/react-native-jsi-sqlite)), but this morning I was trying to write a Hybrid class and the JNI binding code, I got as far as having the JNI code compiling but then I failed to pass the pointers/references to the common CPP library

You can check out the code in the other PR: https://github.com/ospfranco/react-native-quick-sqlite/pull/29

Reply:
@EduFrazao your code looks a lot simpler, would you mind opening a PR so I can review it?

Reply:
> @EduFrazao your code looks a lot simpler, would you mind opening a PR so I can review it?

Sure. I will do it right now. In this moment I'm testing the loadSQLFileAsync method. I will submit it in few minutes and let you know if its working. I will test with several files, some of them with more than 120k rows.

Reply:
well... I got android to compile and I got the call to the callInvoker working, but somehow android it is still crashing. Just open your PR asap, I would like to understand how did you got it working

Reply:
> @EduFrazao your code looks a lot simpler, would you mind opening a PR so I can review it?

Finished my testes here. Async loading of SQLFiles is working, and the applicatoins runs so smooth :).
More than 250k rows background processed, and the UI keeps responsive :).

I will open a PR.

Reply:
I tried to replicate your changes with a struct but somehow android is crashing for me when I try to register theJSI bindings

Reply:
> I tried to replicate your changes with a struct but somehow android is crashing for me when I try to register theJSI bindings

Can I help some how? It fails registering the methods on the sqlite module? I've made tests with the SQLFileLoader and a test method that does nothing, except create some random numbers, put the working thread to sleep (to simulate some workload time).